forge/csrc/autograd/autograd.cpp

```diff
@@ -719,6 +721,152 @@ NodeContext autograd_engine::create_input(
+            fwd_node->name() + "_recompute", fwd_node->as<graphlib::OpNode>()->op_type()),
+        graph->get_subgraph_id_for_node(fwd_node->id()));
+
+    recompute_node->set_shape(fwd_node->shape());
+    recompute_node->set_output_df(fwd_node->output_df());
+    recompute_node->set_epoch_type(graphlib::NodeEpochType::Backward);
+
+    std::unordered_map<graphlib::NodeId, graphlib::NodeId> producer_remap;
+
+    fwd_to_recompute[fwd_node] = recompute_node;
+
+    // Hook up Recompute Node to its operands
+    for (const Edge &edge : graph->operand_data_edges(fwd_node))
+    {
+        Node *fwd_operand = graph->node_by_id(edge.producer_node_id);
+        bool same_epoch = (fwd_operand->is_forward() && fwd_operand->node_type() != graphlib::NodeType::kInput);
+        graphlib::NodeId producer_id = same_epoch ? fwd_to_recompute.at(fwd_operand)->id() : fwd_operand->id();
+
+        if (producer_remap.find(producer_id) != producer_remap.end())
+        {
+            producer_id = producer_remap.at(producer_id);
+        }
+
+        Edge recompute_edge = Edge(
+            producer_id,
+            edge.producer_output_port_id,
+            recompute_node->id(),
+            edge.consumer_input_port_id,
+            edge.edge_type);
+
+        graph->add_edge(recompute_edge);
+
+        auto fwd_edge_attr = graph->get_edge_attributes(edge);
+        auto recompute_edge_attr = graph->get_edge_attributes(recompute_edge);
+        recompute_edge_attr->copy_from(*fwd_edge_attr);
+    }
+
+    return recompute_node;
+}
+
+static void add_edges_for_recompute_op(
+    graphlib::Graph *graph, const graphlib::Edge &fwd_to_bwd_edge, std::unordered_map<Node *, Node *> &fwd_to_recompute)
+{
+    Node *fwd_node = graph->node_by_id(fwd_to_bwd_edge.producer_node_id);
+
+    if (fwd_to_recompute.find(fwd_node) == fwd_to_recompute.end())
+    {
+        return;
+    }
+
+    Node *recompute_node = fwd_to_recompute.at(fwd_node);
+
+    // Replace the original FWD->BWD edge with RECOMPUTE->BWD instead.
+    graph->add_edge(
+        Edge(
+            recompute_node->id(),
+            fwd_to_bwd_edge.producer_output_port_id,
+            fwd_to_bwd_edge.consumer_node_id,
+            fwd_to_bwd_edge.consumer_input_port_id,
+            fwd_to_bwd_edge.edge_type),
+        graph->get_edge_attributes(fwd_to_bwd_edge));
+    graph->remove_edge(fwd_to_bwd_edge);
+}
+
+void autograd_engine::insert_recompute_ops()
+{
+    if (!config.recompute)
+    {
+        return;
+    }
+
+    std::unordered_map<Node *, Node *> forward_to_recompute;
+    std::vector<Node *> topo_order = graphlib::topological_sort(*graph);
+
+    // We loop through only the backward nodes and look for operand nodes that are marked
+    // as FWD. Each instance is an opportunity to perform a potential recompute on the original
+    // FWD op.
+    //
+    // TODO(#395): extend this to support configurable op-specific recompute; rather than an all-or-nothing sort-of
```

Comment:
Like it! Maybe we can create one more issue to try to figure out "automagically" which part of the graph should be recomputed? E.g. based on potential DRAM vs Compute usage (explorational, maybe even for internship :))

Reply:
I didn't want to create additional issues, since at the moment it is not clear whether we want to pursue this path. This is just for note keeping/future work.

I have added a comment in the already created issue #395, to consider automatic recompute as a next step...

Reply:
What @nvukobratTT is definitely a long-term plan which we are aware of. I would agree with @pilkicTT to not open an issue at this point to avoid confusion regarding issue prioritization.